### PR TITLE
Allow Exclusion of Input IDs from RepetitionPenaltyLogitsProcessor

### LINKED ISSUE #36642: Is it correct that the repetition penalty is applied to the input_ids encompassing all inputs and outputs, rather than solely on the generated tokens?
### System Info

- `transformers` version: 4.44.2
- Platform: Linux-3.10.0-1160.el7.x86_64-x86_64-with-glibc2.17
- Python version: 3.10.14
- Huggingface_hub version: 0.26.2
- Safetensors version: 0.4.5
- Accelerate version: 1.0.1
- Accelerate config:    not found
- PyTorch version (GPU?): 2.5.1+cu124 (True)

### Who can help?

@gante 

### Information

- [x] The official example scripts
- [x] My own modified scripts

### Tasks

- [ ] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [x] My own task or dataset (give details below)

### Reproduction

https://github.com/huggingface/transformers/blob/c4d4e8bdbd25d9463d41de6398940329c89b7fb6/src/transformers/generation_utils.py#L101    

    def enforce_repetition_penalty_(self, lprobs, batch_size, num_beams, prev_output_tokens, repetition_penalty):
        """repetition penalty (from CTRL paper https://arxiv.org/abs/1909.05858). """
        for i in range(batch_size * num_beams):
            for previous_token in set(prev_output_tokens[i].tolist()):
                # if score < 0 then repetition penalty has to multiplied to reduce the previous token probability
                if lprobs[i, previous_token] < 0:
                    lprobs[i, previous_token] *= repetition_penalty
                else:
                    lprobs[i, previous_token] /= repetition_penalty

    def postprocess_next_token_scores(
        self,
        scores,
        input_ids,
        no_repeat_ngram_size,
        bad_words_ids,
        cur_len,
        min_length,
        max_length,
        eos_token_id,
        repetition_penalty,
        batch_size,
        num_beams,
    ):
        # repetition penalty (from CTRL paper https://arxiv.org/abs/1909.05858)
        if repetition_penalty != 1.0:
            self.enforce_repetition_penalty_(
                scores, batch_size, num_beams, input_ids, repetition_penalty,
            )
    ...

### Expected behavior

Maybe it should just rely on generated tokens?

### PR DESCRIPTION
**Pull request by alex-jw-brooks**
This PR adds a flag for allowing the exclusion of input IDs when using the `RepetitionPenaltyLogitsProcessor` - currently there are some workarounds that may be used for language models, e.g., passing input embeddings instead of input ids, however, these workarounds are a bit trickier with multimodal models, since we generally pass the input ids first to create the merged multimodal embeddings in the model.

Adding such a flag would be really helpful for models like granite speech, which @avihu111 and I had recently added support for. In experiments, the model works very well with a super high repetition penalty on just the newly generated token IDs, but performance degrades severely when the input IDs are included.

By default, input IDs are included to avoid changing existing default behaviors.

Fixes https://github.com/huggingface/transformers/issues/36642 

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [x] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [x] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

Probably @gante @zucchini-nlp @eustlb would have the most context!

### CODE REVIEW

**Comment by gante** [comment]
Hi Alex! 👋 

This is the first time I'm reading about this feature request. As such, there are a few workarounds I'd like to ask you to try before adding your proposal -- minimizing flags greatly improves the UX of the library in the long run 🤗 If none of them solves your problem (or if they have unwanted secondary effects), let me know! 

_______________________________
1. If we combine `EncoderRepetitionPenaltyLogitsProcessor` (modifier on the prompt tokens) with `RepetitionPenaltyLogitsProcessor` (modifier on all tokens) with the right penalty values, we can cancel out the modifier applied to the prompt. The difference is what happens if a token that is in the prompt gets generated: with the existing solution, that token will never get a modifier, while with your solution it gets a modifier.
2. If we run a forward pass with the prompt to extract the KV cache, and then call generate passing the KV cache without passing the prompt tokens, the penalty only gets applied to new tokens. The caveat here is that you need to generate the next token from the prompt manually after the forward pass, and feed it to `generate` as input (so generate receives the cache + 1st generated token). You'll also need to hand-craft the attention mask for `generate`. If this works and it's not too complex on your side, I can open a PR to automate the creation of the attention mask when partial input tokens + KV cache are passed to `generate`.

**Comment by alex-jw-brooks** [comment]
Hi @gante, thanks a lot for the quick response! That all makes total sense and I agree. I'm not a fan of the flag either, but thought it might be nice to avoid changing the default behavior.

Happy to look into your suggestions, it's greatly appreciated! Do you have any thoughts on keeping the change to the `RepetitionPenaltyLogitsProcessor`, but just avoiding exposing it through`.generate`? This might be simpler, because then if people do end up hitting this in the future, there is a clear workaround by passing it as a custom logit processor, and it avoids touching the generate api 🙂  


```python
from transformers import RepetitionPenaltyLogitsProcessor

logits_processor = [
    RepetitionPenaltyLogitsProcessor(
        penalty=3.0,
        input_ids_seq_length=model_inputs["input_ids"].shape[-1],
    )
]

model_outputs = model.generate(
    **model_inputs,
    ...
    logits_processor=logits_processor,
)
```

**Comment by gante** [comment]
Yeah, I'd be happy to add the flag to the processor, but not to the generation config -- it would indeed be the best of both worlds 🤗

File: src/transformers/generation/logits_process.py [hunk 1]
```diff
@@ -303,6 +304,8 @@ class RepetitionPenaltyLogitsProcessor(LogitsProcessor):
         penalty (`float`):
             The parameter for repetition penalty. 1.0 means no penalty. Above 1.0 penalizes previously generated
             tokens. Between 0.0 and 1.0 rewards previously generated tokens.
+        input_ids_seq_length (`int`, *optional*, defaults to 0):
```
Comment (gante) [inline src/transformers/generation/logits_process.py:307]: Let's rename it to something like `prompt_ignore_length`, so that it is clear that it modifies the original behavior.

`input_ids_seq_length`, at a first glance, looks like an input we should provide to get the correct behavior, as opposed to a modified behavior.

Let's also document the flag accordingly (it's an optional modifier), and explain that if it is to be used the class must be passed as a custom processor to generate. Ideally, with an example in the doctest below :D

File: src/transformers/generation/logits_process.py [hunk 2]
```diff
@@ -326,14 +329,18 @@ class RepetitionPenaltyLogitsProcessor(LogitsProcessor):
     ```
     """
 
-    def __init__(self, penalty: float):
+    def __init__(self, penalty: float, input_ids_seq_length: Optional[int] = 0):
```
Comment (gante) [inline src/transformers/generation/logits_process.py:332]:
```suggestion
    def __init__(self, penalty: float, input_ids_seq_length: Optional[int] = None):
```
We prefer to default optional flags to `None` whenever possible and sensible. That way we can distinguish between no user input and user input = 0, which may be handy in the future 🤗

**Review by gante — COMMENTED** [review]
Looks mostly good to me, added a few documentation/signature nits 🤗

**Review by gante — APPROVED** [review]
LGTM, thank you for iterating 🤗

**Comment by alex-jw-brooks** [comment]
Awesome, thanks a lot for the quick reviews! 🙂
